document minumum replicas constraint (#855)
elastisys/compliantkubernetes: docs/user-guide/safeguards/enforce-minimum-replicas.md

---
tags: []
---

# Avoid Downtime with Replicas

Compliant Kubernetes by default recommends a minimum of 2 replicas for Deployments (and StatefulSets).

Therefore a warning will be issued when the number of replicas is less than 2.

```console
[elastisys-warn-minimum-replicas] The provided number of replicas is low for Deployment: demo-ck8s-user-demo. Elastisys Compliant Kubernetes recommends a minimum of 2 replicas.
```

You can either resolve this by setting the number of replicas to a minimum of 2.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: user-demo
spec:
  replicas: 2
```

> Or in the case where the deployment is not of high priority you can suppress the warning by adding a label.

Reviewer: I think this might be the case for some people as well. Do you think it makes sense to add this as well? Feel free to reword it if you want.
```suggestion
Or in the case where the deployment is not of high priority or there are technical limitations preventing you from using more than one replica you can suppress the warning by adding a label.
```

> Or in the case where the deployment is not of high priority or there are technical limitations preventing you from using more than one replica you can suppress the warning by adding a label.

Author: Yeah commas definetly make it look better.
It's not capitalized as i wanted to say both 'Deployment' and 'StatefulSet', but i guess could just mention them both.

> Or, in the case where the Deployment or StatefulSet is not of high priority or there are technical limitations preventing you from using more than one replica, you can suppress the warning by adding a label.

Reviewer: ```suggestion
Or, in the case where the Deployment or StatefulSet is not of high priority or there are technical limitations preventing you from using more than one replica, you can suppress the warning by adding an annotation.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: user-demo

>   labels:

Reviewer 2: ```suggestion
  annotations:
```

Annotations are preferred for non-identifying metadata about Kubernetes objects: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

>     ignoreMinimumReplicas: "yes" # The value part can be anything.

Reviewer 2: ```suggestion
    elastisys.io/ignore-minimum-replicas: "yes" # The value part can be anything.
```

Please prefix this with `elastisys.io` to make it clear that this is in a "label namespace" that we own. Also, the naming convention for annotations seems to be lowercase-dashes.
